sftp client

pellcorp opened the pull request:
sftp storage support

pellcorp commented:
The tests for sftp only run when used with vagrant at the moment, because there needs to be sftp infrastructure configured, I just followed the approach for s3, etc which are skipped by default.  Also due to https://github.com/paramiko/paramiko/issues/1369, I have pinned cryptogrophy to 2.4.2, not sure if this is ok or not.

rikonen commented on Vagrantfile:
```diff
@@ -2,34 +2,45 @@
+        sed -i 's/PasswordAuthentication no/PasswordAuthentication yes/g' /etc/ssh/sshd_config
+        systemctl reload ssh
+        useradd -m -U u12345678
+        echo "u12345678:password" | chpasswd
```
I'm not sure how the networking is set up and can you actually make external connections but this would in any event look less suspicious if password was derived from urandom and stored in some file from which it can be read locally in tests.

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+        self.password = password
+        self.private_key = private_key
+
+        if password is None and private_key is None:
```
`if not password and not private_key:`? I'm guessing empty string isn't better than None.

pellcorp replied:
This would mean that the sftp server supports empty passwords, which I guess is possible, I will check.

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+        self.log.debug("Get file content: %r", target_path)
+
+        try:
+            return self.client.getfo(target_path, fileobj_to_store_to, progress_callback)
```
Maybe add comment here mentioning that pysftp conveniently happens to have the exact same signature for progress callback as pghoard. This looks suspicious otherwise.

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+import logging
+import warnings
+import pysftp
+from pysftp import CnOpts
```
This import looks useless, could just use `pysftp.CnOpts` in the code later.

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+import datetime
+from stat import S_ISDIR
+from io import BytesIO, StringIO
+import json
```
The import ordering doesn't seem to be all that consistent in existing code either but mostly it would seem to be `from` imports first followed by `import` imports, both sorted alphabetically within the group. That would look nicer than the current mix of imports used here.

rikonen commented on requirements.txt:
```diff
@@ -1,6 +1,6 @@
 astroid==2.0.0
 botocore 
-cryptography 
+cryptography=2.4.2
```
Does it need to be exactly this or would `>=` work?

pellcorp replied:
It does because of a issue with cryptography 5.0.0: https://github.com/paramiko/paramiko/issues/1369

pellcorp replied:
I will do a bit more playing around and see if I can still get it working with >= 2.4.2 though

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+        target_path = self.format_key_for_backend(key.strip("/"))
+        self.log.info("Removing path: %r", target_path)
+        try:
+            self.client.remove(target_path)
```
Must delete metadata as well

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+
+        self.client.makedirs(os.path.dirname(target_path))
+        self.client.putfo(flo=fd, remotepath=target_path, callback=wrapper_upload_progress_fn)
+        self._save_metadata(target_path, metadata)
```
Saving metadata first would prevent running into race conditions when listing files where you get back files that have contents but no metadata.

pellcorp replied:
The metadata is also the flag file and if a data file has no flag it's not returned in the list thus why it's written only after the data file. I copied this idea from the local provider, this affects some of your comments above too, but when I get to a computer I will review in more detail to ensure that this is working as I intended.

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+                upload_progress_fn(bytes_written)
+
+        self.client.makedirs(os.path.dirname(target_path))
+        self.client.putfo(flo=fd, remotepath=target_path, callback=wrapper_upload_progress_fn)
```
I didn't check how this is actually implemented but storing the file should be done in somehow atomic manner so that listing files does not return files that are not yet fully uploaded. If this is automatically handled then that's fine, if not then presumably you should first upload with some temp name that is ignored from listing and then do rename once the upload is complete.

pellcorp replied:
I use the metadata file to exclude files currently being copied.  The metadata file should be small enough its written in a single block, no better way that I can see with sftp protocol

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+
+        files = []
+        for attr in attrs:
+            files.append(attr.filename)
```
Could use comprehension and since this is used up for doing `in` checks using `set` would work much better. `files = set(attr.filename for attr in attrs)`

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+                    return
+                except FileNotFoundError:
+                    return
+
```
Extra linefeed

rikonen commented on pghoard/rohmu/object_storage/sftp.py:
```diff
@@ -0,0 +1,209 @@
+                            "metadata": metadata,
+                        }
+                    )
+                else:
```
Doesn't seem to have actual effect.

rikonen commented on test/test_storage.py:
```diff
@@ -81,11 +81,13 @@ def progress_callback(pos, total):
-    assert st.get_contents_to_string("test_copy/copy1") == (b"dummy", {"k": "v"})
-    st.copy_file(source_key="test1/x1", destination_key="test_copy/copy2", metadata={"new": "meta"})
-    assert st.get_contents_to_string("test_copy/copy2") == (b"dummy", {"new": "meta"})
+    # sftp does not support remote copt
```
Typo (copt)

rikonen commented on test/test_storage.py:
```diff
@@ -137,11 +139,15 @@ def progress_callback(pos, total):
         created_keys.add("test1/sub3/sub3.1/sub3.1.1")
 
-    if driver == "local":
+    if driver in ("local", "sftp"):
```
Maybe list or set instead (same for other similar checks)? Tuples are nasty when you remove all but the last element and don't remember the extra comma.

pellcorp commented:
OK, fixes applied based on your feed back.

pellcorp commented:
and build has passed :-)

rikonen commented on test/test_storage.py:
```diff
@@ -312,17 +318,26 @@ def upload_progress(progress):
+        # for no prefix testing, we need to cleanup existing files
+        os.system("sudo rm -rf /home/u12345678/*")
+
+        with open('/etc/sftp-password', 'r') as pwdfile:
```
Nitpick: Looks like `"` should be preferred over `'`. Also, I'd reverse the order so that `rm -rf` is done only after successfully reading the password to make it even more unlikely this ends up blowing away some poor `u12345678` user's data on a machine that happened to have `/etc/vagrantvm` file.

mteiste commented:
Not sure if this relevant, but in the past I've had issues with paramiko's sftp implementation when transferring large files; if a re-key happened while the file transfer was ongoing the transfer failed.

rikonen commented:
> when transferring large files

"large" here apparently being something between 10 and 100 MiB. @pellcorp if you didn't already test with some sizable (typically basebackup chunks can be anything up to a gigabyte or two) files maybe you could do some test with those to check if that issue is still relevant.

pellcorp commented:
We have tested downloads using this code to 30+ gb files, and I just confirmed that  a 1gb upload worked just fine.  Will make the other changes.

pellcorp commented:
Good point about the /tmp/vagrantvm file existing, I will clean that up

pellcorp commented:
ok I changed to generate a random username and password to avoid clobbering an existing user, also I cleaned up the flag file stuff and put it into the /home/vagrant directory.

pellcorp commented:
build has passed :-)

rikonen commented on test/test_storage.py:
```diff
@@ -312,17 +318,26 @@ def upload_progress(progress):
+        storage_config = {'server': 'localhost', 'port': 22, 'username': username, 'password': password}
+
+        # for no prefix testing, we need to cleanup existing files
+        os.system('sudo rm -rf /home/' + username + '/*')
```
Couple of nitpicks: 1) `"` is used all around this new block so that should be used instead of `'` and `.format()` is nicer than `+`.

And more importantly: I would add something that bails out in case `username` is empty or else really bad things may happen here...

pellcorp replied:
ok, I changed the whole file to use single quotes, even though most of it I did not touch, and I won't rm if username is not defined.

rikonen replied:
Sorry you misunderstood my comment. I meant this block should be using double quotes as well because that's what's predominantly used in pghoard codebase otherwise.

pellcorp replied:
ok will change it back and get rid of the additional commit

pellcorp commented:
ok build passed

pellcorp commented:
hopefully this time I am done :-)

rikonen commented on test/test_storage.py:
```diff
@@ -312,17 +318,27 @@ def upload_progress(progress):
         storage_config = conf_func()
 
-    if storage_type in ("aws_s3", "ceph_s3"):
+    if storage_type in ["aws_s3", "'ceph_s3"]:
```
Extra `'` before `ceph_s3`

pellcorp commented:
ok lets try now :-)

rikonen commented:
@pellcorp I didn't do proper enough due diligence check for the pysftp library used here. It seems the library isn't maintained anymore and isn't available on Fedora so this breaks things for all Fedora users. To unbreak things we decided that the best course of action for now is to revert this change (https://github.com/aiven/pghoard/pull/340) and then see how to make this in a way that works for Fedora as well.

Because pysftp seems to be such a lightweight wrapper our thinking is that it doesn't make sense to try to set up things so that it would be directly usable on Fedora somehow but instead just implement things directly using paramiko. Is that something you could look into?

pellcorp commented:
I will have a go when I have time, a shame as I was hopeful I had finished this work

pellcorp commented:
I am looking at this now, hopeful to have something soonish

pellcorp commented:
I have opened a new merge request with sftp client using paramiko directly

https://github.com/aiven/pghoard/pull/342
